temporal/lite_server.go

```diff
@@ -0,0 +1,379 @@
+		VisibilityStore:  sqliteplugin.PluginName,
+		NumHistoryShards: 1,
+		DataStores: map[string]config.DataStore{
+			sqliteplugin.PluginName: {SQL: &sqliteConfig},
+		},
+	}
+	serverConfig.ClusterMetadata = &cluster.Config{
+		EnableGlobalNamespace:    false,
+		FailoverVersionIncrement: 10,
+		MasterClusterName:        "active",
+		CurrentClusterName:       "active",
+		ClusterInformation: map[string]cluster.ClusterInformation{
+			"active": {
+				Enabled:                true,
+				InitialFailoverVersion: 1,
+				RPCAddress:             fmt.Sprintf("%s:%d", localBroadcastAddress, cfg.FrontendPort),
+			},
+		},
+	}
+	serverConfig.DCRedirectionPolicy = config.DCRedirectionPolicy{
+		Policy: "noop",
+	}
+	serverConfig.Services = map[string]config.Service{
+		"frontend": cfg.mustGetService(0, provider),
+		"history":  cfg.mustGetService(1, provider),
+		"matching": cfg.mustGetService(2, provider),
+		"worker":   cfg.mustGetService(3, provider),
+	}
+	serverConfig.Archival = config.Archival{
+		History: config.HistoryArchival{
+			State:      "disabled",
+			EnableRead: false,
+			Provider:   nil,
+		},
+		Visibility: config.VisibilityArchival{
+			State:      "disabled",
+			EnableRead: false,
+			Provider:   nil,
+		},
+	}
+	serverConfig.PublicClient = config.PublicClient{
+		HostPort: fmt.Sprintf("%s:%d", localBroadcastAddress, cfg.FrontendPort),
+	}
+	serverConfig.NamespaceDefaults = config.NamespaceDefaults{
+		Archival: config.ArchivalNamespaceDefaults{
+			History: config.HistoryArchivalNamespaceDefaults{
+				State: "disabled",
+			},
+			Visibility: config.VisibilityArchivalNamespaceDefaults{
+				State: "disabled",
+			},
+		},
+	}
+}
+
+func (cfg *LiteServerConfig) applyDefaults() {
+	if cfg.BaseConfig == nil {
+		cfg.BaseConfig = &config.Config{}
+	}
+	if cfg.Logger == nil {
+		cfg.Logger = log.NewZapLogger(log.BuildZapLogger(log.Config{
+			Stdout:     true,
+			Level:      "info",
+			OutputFile: "",
+		}))
+	}
+}
+
+func (cfg *LiteServerConfig) validate() error {
+	for pragma := range cfg.SQLitePragmas {
+		if _, ok := supportedPragmas[strings.ToLower(pragma)]; !ok {
+			return fmt.Errorf("unsupported SQLite pragma %q. allowed pragmas: %v", pragma, getAllowedPragmas())
+		}
+	}
+
+	if cfg.Ephemeral && cfg.DatabaseFilePath != "" {
+		return fmt.Errorf("config option DatabaseFilePath is not supported in ephemeral mode")
+	}
+	if !cfg.Ephemeral && cfg.DatabaseFilePath == "" {
+		return fmt.Errorf("config option DatabaseFilePath is required when ephemeral mode disabled")
+	}
+
+	return nil
+}
+
+// LiteServer is a high level wrapper for temporal.LiteServer that automatically configures a SQLite backend.
+type LiteServer struct {
+	internal         Server
+	frontendHostPort string
+}
+
+// NewLiteServer initializes a Server with a SQLite backend.
+//
+// If the db file does not already exist, schema migrations are automatically run.
+func NewLiteServer(liteConfig *LiteServerConfig, opts ...ServerOption) (*LiteServer, error) {
```

Comment:
Yeah, I was also hoping it could hook into the existing ServerOptions more deeply, and then maybe add the new methods to Server itself. Of the new methods: Start/Stop just pass through, NewClient* can be modified to use membership (the same way the worker does it), the only tricky part is FrontendHostPort, but we know what port/address we bound to, we just have to refactor the code a little to make it available.

Some of the Lite options are applicable to the server too, e.g. namespaces to auto-register.

In theory I like the idea of a smooth continuum between a "lite" configuration and a "real" server, and it seems pretty easy to do that without compromising the ease of use of "lite", which I know is a priority. (Just wrap the LiteConfig in a ServerOption.) But I admit it might not provide much value beyond the theory, and we might be better served by two entry points. I'll think about this a little more but would appreciate any arguments either way.

Reply:
> slightly hacky "run sqlite migrations, create namespaces, register search attributes" stuff

When you write it like that, it sounds a lot like what auto-setup is doing. I wonder if this is a potential path towards redoing auto-setup in a nicer way and deprecating the existing (also hacky) shell script… @alexshtin